python/cog/server/worker.py

```diff
@@ -163,9 +161,57 @@ def predict(
             self._assert_state(WorkerState.READY)
             result = Future()
             self._predictions_in_flight[tag] = PredictionState(tag, payload, result)
-        self._request_send_conn.send(PredictionRequest(tag))
+
+        self._prediction_start_pool.submit(self._start_prediction(tag, payload))
         return result
 
+    def _start_prediction(
+        self, tag: Optional[str], payload: Dict[str, Any]
+    ) -> Callable[[], None]:
+        def start_prediction() -> None:
+            try:
+                to_await = []
+                futs = {}
+                # Prepare payload asynchronously (download URLPath objects)
+                for k, v in payload.items():
+                    # Check if v is an instance of URLPath
+                    if isinstance(v, URLPath):
+                        futs[k] = self._input_download_pool.submit(v.convert)
+                        to_await.append(futs[k])
+                    # Check if v is a list of URLPath instances
+                    elif isinstance(v, list) and all(
+                        isinstance(item, URLPath) for item in v
+                    ):
+                        futs[k] = [
+                            self._input_download_pool.submit(item.convert) for item in v
+                        ]
+                        to_await += futs[k]
+                futures.wait(to_await, return_when=futures.FIRST_EXCEPTION)
```

Comment:
What happens to the rest of the work if we exit on first failure? Do we need to clean them up?

Reply:
i suppose we could go cancel the unfinished futures. I'll do that.